op-node/rollup/interop/managed/attributes.go

```diff
@@ -0,0 +1,112 @@
+	}
+	// Add the system-tx that declares the replacement.
+	l2Output := eth.OutputV0{
+		StateRoot:                invalidatedBlock.ExecutionPayload.StateRoot,
+		MessagePasserStorageRoot: eth.Bytes32{}, // TODO: the withdrawals-root in header is needed here.
+		BlockHash:                invalidatedBlock.ExecutionPayload.BlockHash,
+	}
+	outputRootPreimage := l2Output.Marshal()
+	invalidatedBlockTx := InvalidatedBlockSourceDepositTx(outputRootPreimage)
+	invalidatedBlockOpaqueTx, err := invalidatedBlockTx.MarshalBinary()
+	if err != nil {
+		panic(fmt.Errorf("failed to encode system-deposit: %w", err))
+	}
+	replaceTxs = append(replaceTxs, invalidatedBlockOpaqueTx)
+
+	// Optional engine-API attribute, because of L1-compat, and thus we need a pointer to it.
+	gasLimit := invalidatedBlock.ExecutionPayload.GasLimit
+
+	// unfortunately, the engine API needs the inner value, not the extra-data.
+	// So we translate it here.
+	extraData := invalidatedBlock.ExecutionPayload.ExtraData
+	denominator, elasticity := eip1559.DecodeHoloceneExtraData(extraData)
+	eip1559Params := eth.Bytes8(eip1559.EncodeHolocene1559Params(denominator, elasticity))
+
+	attrs := &eth.PayloadAttributes{
+		Timestamp:             invalidatedBlock.ExecutionPayload.Timestamp,
+		PrevRandao:            invalidatedBlock.ExecutionPayload.PrevRandao,
+		SuggestedFeeRecipient: invalidatedBlock.ExecutionPayload.FeeRecipient,
+		Withdrawals:           invalidatedBlock.ExecutionPayload.Withdrawals,
+		ParentBeaconBlockRoot: invalidatedBlock.ParentBeaconBlockRoot,
+		Transactions:          replaceTxs,
+		NoTxPool:              true,
+		GasLimit:              &gasLimit,
+		EIP1559Params:         &eip1559Params,
+	}
+	return attrs
+}
+
+func InvalidatedBlockSourceDepositTx(outputRootPreimage []byte) *types.Transaction {
+	outputRoot := crypto.Keccak256Hash(outputRootPreimage)
+	src := derive.InvalidatedBlockSource{OutputRoot: outputRoot}
+	return types.NewTx(&types.DepositTx{
+		SourceHash:          src.SourceHash(),
+		From:                derive.L1InfoDepositerAddress,
+		To:                  &common.Address{}, // to the zero address, no EVM execution.
+		Mint:                big.NewInt(0),
+		Value:               big.NewInt(0),
+		Gas:                 36_000,
+		IsSystemTransaction: false,
+		Data:                outputRootPreimage,
+	})
+}
+
+func DecodeInvalidatedBlockTxFromReplacement(txs []eth.Data) (*eth.OutputV0, error) {
+	if len(txs) == 0 {
+		return nil, errors.New("expected block-replacement tx and more")
+	}
+	var tx types.Transaction
+	if err := tx.UnmarshalBinary(txs[len(txs)-1]); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal invalidated-block system tx: %w", err)
+	}
+	return DecodeInvalidatedBlockTx(&tx)
+}
+
+func DecodeInvalidatedBlockTx(tx *types.Transaction) (*eth.OutputV0, error) {
+	if tx.Type() != types.DepositTxType {
+		return nil, fmt.Errorf("expected deposit tx type, but got %d", tx.Type())
+	}
+	signer := types.LatestSignerForChainID(tx.ChainId())
```

Comment:
Is it necessary to validate `tx.ChainId()` matches the actual chainid here?

Reply:
It's implicitly checked below when recovering the sender from the tx.